src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs

```diff
@@ -0,0 +1,212 @@
+    {
+        public static ImmutableArray<TResult>
+            GetImportanceMetricsMatrix(
+                IHostEnvironment env,
+                ITransformer model,
+                IDataView data,
+                Func<IDataView, TResult> evaluationFunc,
+                Func<TResult, TResult, TResult> deltaFunc,
+                string features,
+                int topExamples)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            var host = env.Register(nameof(PermutationFeatureImportance<TResult>));
+            host.CheckValue(model, nameof(model));
+            host.CheckValue(data, nameof(data));
+            host.CheckValue(features, nameof(features));
+
+            VBuffer<ReadOnlyMemory<char>> slotNames = default;
+            var metricsDelta = new List<TResult>();
+            var progressIterations = 10;
+
+            using (var ch = host.Start("GetImportanceMetrics"))
+            {
+                ch.Trace("Scoring and evaluating baseline.");
+                var baselineMetrics = evaluationFunc(model.Transform(data));
+
+                // Get slot names.
+                var featuresColumn = data.Schema[features];
+                int numSlots = featuresColumn.Type.VectorSize;
+                data.Schema.TryGetColumnIndex(features, out int featuresColumnIndex);
+
+                ch.Info("Number of slots: " + numSlots);
+                if (data.Schema.HasSlotNames(featuresColumnIndex, numSlots))
+                    data.Schema.GetMetadata(MetadataUtils.Kinds.SlotNames, featuresColumnIndex, ref slotNames);
+
+                if (slotNames.Length != numSlots)
+                    slotNames = VBufferUtils.CreateEmpty<ReadOnlyMemory<char>>(numSlots);
+
+                var workingFeatureIndices = Enumerable.Range(0, numSlots).ToList();
+
+                // Note: this will not work on the huge dataset.
+                var maxSize = topExamples > 0 ? topExamples : Utils.ArrayMaxSize;
+                List<float> initialfeatureValuesList = new List<float>();
+
+                // Cursor through the data to cache slot 0 values for the upcoming permutation.
+                var valuesRowCount = 0;
+                // REVIEW: Seems like if the labels are NaN, so that all metrics are NaN, this command will be useless.
+                // In which case probably erroring out is probably the most useful thing.
+                using (var cursor = data.GetRowCursor(col => col == featuresColumnIndex))
+                {
+                    var featuresGetter = cursor.GetGetter<VBuffer<float>>(featuresColumnIndex);
+                    var featuresBuffer = default(VBuffer<float>);
+
+                    while (initialfeatureValuesList.Count < maxSize && cursor.MoveNext())
+                    {
+                        featuresGetter(ref featuresBuffer);
+                        initialfeatureValuesList.Add(featuresBuffer.GetItemOrDefault(workingFeatureIndices[0]));
```

Comment:
>workingFeatureIndices[0] [](start = 85, length = 24)

this looks suspicious.
`workingFeatureIndices[0]` is always 0, why we need `workingFeatureIndices[0]`

Reply:
Honestly I don't understand why we even need `workingFeatureIndices` glancing through code, we never modify it, and it just array with values from 0 to numSlots.

---
In reply to: [235190185](https://github.com/dotnet/machinelearning/pull/1658#discussion_r235190185) [](ancestors = 235190185)

Reply:
OK, I've looked into internal repo, and we have one specific case for the model with weights and we populate that array differently for that case.
Why it got dropped?

---
In reply to: [235190701](https://github.com/dotnet/machinelearning/pull/1658#discussion_r235190701) [](ancestors = 235190701,235190185)

Reply:
Here is the comment from Shahab: 
* Weight Filtering: TLC had weight filtering features by using IPredictorWithFeatureWeights. We can't do this in ML.NET so I removed that option.


---
In reply to: [235195353](https://github.com/dotnet/machinelearning/pull/1658#discussion_r235195353) [](ancestors = 235195353,235190701,235190185)